content/article/deploying-dancer-apps-addendum.md

```diff
@@ -0,0 +1,164 @@
+rewrote the “program_args” code to be this:
+
+```perl
+program_args => [ '--workers', ($ENV{KLORTHO_APP_WORKERS} // 10),
+                  '-l', ":$ENV{KLORTHO_APP_PORT}",
+                  dirname(abs_path($0)) . '/app.psgi' ],
+```
+
+I made similar changes to all the “app_service” files, added appropriate
+environment variables to all the “.env” files and restarted all the
+apps. Immediately, I could see an improvement as I was now running maybe
+a third of the app processes on the server. But I felt I could do better.
+So I had a close look at the Starman documentation to see if there was
+anything else I could tweak. That’s when I found the “–preload-app”
+command-line option.
+
+Starman works by loading a main driver process which then fires up as many
+worker processes as you ask it for. Without the “–preload-app” option,
+each of those processes loads a copy of the application. But with this
+option, each worker processes reads the main driver’s copy of the
+application and only loads its own copy when it wants to write something.
+This can be a big memory saving – although it’s important to note that the
+documentation warns:
+
+> Enabling this option can cause bad things happen when resources like sockets
+> or database connections are opened at load time by the master process and
+> shared by multiple children.
+
+I’m pretty sure that most of my apps are not in any danger here, but I’m
+keeping a close eye on the situation and if I see any problems, it’s easy
+enough to turn preloading off again.
+
+When adding the preloading option to “app_service”, I realised I should
+probably completely rewrite the code that builds the program arguments. It
+now looks like this:
+
+```perl
+my @program_args;
+if ($ENV{KLORTHO_WORKER_COUNT}) {
+  push @program_args, '--workers', $ENV{KLORTHO_WORKER_COUNT};
+}
+if ($ENV{KLORTHO_APP_PORT}) {
+  push @program_args, '-l', ":$ENV{KLORTHO_APP_PORT}";
+}
+if ($ENV{KLORTHO_APP_PRELOAD}) {
+  push @program_args, '--preload-app';
+}
+push @program_args, dirname(abs_path($0)) . '/bin/app.psgi';
+```
+
+The observant among you will notice that I’ve subtly changed the behaviour
+of the worker count environment variable. Previously, a missing variable
+would use a default value of 10. Now, it just omits the argument which
+uses Starman’s default value of 5.
+
+I’ve made similar changes in all my “app_service” programs and set
+environment variables to turn preloading on. And now my apps use
+substantially less memory. The server hasn’t died since I implemented this
+stuff at the start of this week. So that makes me very happy.
+
+But programming is the pursuit of minimisation. I’ve already seen two places
+where I can make these programs smaller and simpler.
+
+1. That last code snippet looks too repetitive. It should be a loop 
+   iterating over a hash. The keys are the names of the environment variables 
+   and the values are references to arrays containing that are added to the 
```

Comment:
Not sure about "to arrays containing". Should we say what they contain?

Reply:
Yes. Fixed.